File: LeetCode/Medium/0200-number-of-islands/0200-number-of-islands.cpp
class Solution {
public:
    int dx[4] = {-1, 0, 1, 0};
    int dy[4] = {0, 1, 0, -1};

    int numIslands(vector<vector<char>>& grid) {
        bool VISIT[305][305] = {};
        int ans = 0;

        int M = grid.size(), N = grid[0].size();

        for(int i = 0; i < M; i++){
            for(int j = 0; j < N; j++){
                if(VISIT[i][j]) continue;
                if(grid[i][j] == '0') continue;
                queue<pair<int, int>> Q;

                VISIT[i][j] = true;       
                ans++;
                Q.push({i, j});

                while(!Q.empty()){
                    auto cur = Q.front();
                    Q.pop();

                    for(int dir = 0; dir < 4; dir++){
                        int nx = cur.first + dx[dir];
                        int ny = cur.second + dy[dir];

                        if(nx < 0 || nx >= M || ny < 0 || ny >= N) continue;
                        if(VISIT[nx][ny]) continue;
                        if(grid[nx][ny] == '0') continue;

                        VISIT[nx][ny] = true;
                        Q.push({nx, ny});
                    }
                }
            }
        }

        return ans;
    }
};

File: LeetCode/Medium/0200-number-of-islands/0200-number-of-islands.java
class Solution {
    int dx[] = {-1, 0, 1, 0};
    int dy[] = {0, 1, 0, -1};

    public int numIslands(char[][] grid) {
        record Grid(int x, int y) {}

        int ans = 0;
        int N = grid.length, M = grid[0].length;

        Queue<Grid> Q = new ArrayDeque<>();
        
        for(int i = 0; i < N; i++){
            for(int j = 0; j < M; j++){
                if(grid[i][j] == '0') continue;

                ans++;

                grid[i][j] = '0';
                Q.offer(new Grid(i, j));

                while(!Q.isEmpty()){
                    Grid cur = Q.poll();
                    int curX = cur.x(), curY = cur.y();

                    for(int dir = 0; dir < 4; dir++){
                        int nxtX = curX + dx[dir];
                        int nxtY = curY + dy[dir];

                        if(nxtX < 0 || nxtX >= N || nxtY < 0 || nxtY >= M) continue;
                        if(grid[nxtX][nxtY] == '0') continue;

                        grid[nxtX][nxtY] = '0';
                        Q.offer(new Grid(nxtX, nxtY));
                    }
                }

            }
        }
        
        return ans;
    }
}